docs/adrs/adr-0001-dependency_management.md

```diff
@@ -0,0 +1,41 @@
+
+## Decision Drivers
+- Minimize the potential for incompatible dependencies
+- Have guardrails/automation so that changes to dependencies made by developers will easily be reflected in the PRs they make
+- Ideally no disruption to user workflow
+- Minimal distruption to developer workflow
+
+## Options
+
+### 1. pip-tools to generate requirements.txt
+Jetstream currently uses this process to manage dependencies.  It involves uses pip-tools to generate a requirements.txt and requirements.in file from the setup.py. [dependabot](https://github.blog/2020-06-01-keep-all-your-packages-up-to-date-with-dependabot/) is a built-in github tool that is used to periodically check for dependency updates and create pull requests to implement them.  [Here](https://github.com/mozilla/jetstream/pull/2017) is an example of such a pull request.
+
+* +: Jetstream uses it already so there is an existing base of experties in the data team
+* +: Dependabot is a built-in github tool that has additional uses like checking for security vulnerabilities
+* +: Dependencies can be automatically updated with a script like the [update_deps](https://github.com/mozilla/jetstream/blob/main/script/update_deps) script in Jetstream
+* -: Still using pip's dependency resolver, which is fast but can have issues with incompatible versions.  This may be mitigated or even solved by the way pip-tools pins dependencies.
+* -: Jetstream doesn't use `pyproject.toml` for configuring dependencies, though it does seem to be supported by `pip-tools`
+
+### 2. poetry
+(poetry)[https://python-poetry.org/] is a popular tool for dependency managment.  When starting development with a poetry project, one uses the `poetry install` command to create a virtual environment for the project and install the package and its dependencies.  New dependencies are added with the `poetry add` command.  poetry (and other tools) are configured in the `pyproject.toml`.
+
+* +: Very reliable dependency resolution
+* +: Uses `pyproject.toml`, which is the new standard and is nice for consolidating other config files
+* +: Also acts as a virtualenv manager
+* +: No changes to user workflow
+* +: When a new dependency is added with `poetry add` the `poetry.lock` and `pyproject.toml` files are automatically updated so there minimal risk of a new dependency not being accounted for
+* -: Dependency resolution is slow, when you add a new package it can take several minutes to figure it out
+* -: Developers would have to modify their workflow to accomodate poetry's virtualenv management
+* -: Not wildely use in the data team
+* -: would need to update tox to use poetry to build and publish package
```

Comment:
I wonder if it makes sense to replace tox if we decide to use poetry?

Reply:
Yeah one annoying thing about poetry is that when you use it with nice CI tools like tox you get the matryoshka doll of virtualenvs.  I'm not sure what the alternative to tox that plays nicely with poetry and all the other moving pieces is though.

Reply:
Well the goal of both of them is to have a consistent environment, so I was thinking that maybe with poetry you don't need tox.

Reply:
Yeah I guess it wouldn't be that hard to take the commands from the `tox.ini` and move them into the circleci config if we were using poetry

Reply:
as a suggestion, have we considered using mamba/conda to generate a virtual environment with all the requirements. it's pretty robust with creating an `environment.yml` and it is quite fast and performant

Reply:
That'd be worth looking into, would you mind adding a section to the ADR for mamba/conda?